src/useTowerGame.ts

```diff
@@ -1,103 +1,130 @@
-      position: [0, 2 * BOX_HEIGHT, 0],
-      args: [START_DIMENSIONS.width, BOX_HEIGHT, START_DIMENSIONS.length],
-      color: 'DarkGreen',
-    },
-  ];
+    const [wasPressed, setWasPressed] = useState(false);
+    const spacePressed = useKeyboardControls<Controls>((state) => state.hit);
+
+    const stackNewBox = useCallback(() => {
+        const lastBoxEdges = new Box3().setFromObject(lastBox.current);
+        const movingBoxEdges = new Box3().setFromObject(movingBox.current);
+        let newWidth = lastBoxEdges.max.x - lastBoxEdges.min.x;
+        let newLength = lastBoxEdges.max.z - lastBoxEdges.min.z;
+        let newStartX = lastBoxEdges.min.x + newWidth / 2;
+        let newStartZ = lastBoxEdges.min.z + newLength / 2;
+
+        if (boxIsOverLappingPos('x', lastBoxEdges, movingBoxEdges)) {
+            newWidth = lastBoxEdges.max.x - movingBoxEdges.min.x;
+            newStartX = movingBoxEdges.min.x + newWidth / 2;
+        } else if (boxIsOverLappingNeg('x', lastBoxEdges, movingBoxEdges)) {
+            newWidth = movingBoxEdges.max.x - lastBoxEdges.min.x;
+            newStartX = lastBoxEdges.min.x + newWidth / 2;
+        } else if (boxIsOverLappingPos('z', lastBoxEdges, movingBoxEdges)) {
+            newLength = lastBoxEdges.max.z - movingBoxEdges.min.z;
+            newStartZ = movingBoxEdges.min.z + newLength / 2;
+        } else if (boxIsOverLappingNeg('z', lastBoxEdges, movingBoxEdges)) {
+            newLength = movingBoxEdges.max.z - lastBoxEdges.min.z;
+            newStartZ = lastBoxEdges.min.z + newLength / 2;
+        } else {
+            dispatch({ type: 'END_GAME' });
+
+            return;
+        }
+
+        const newBox: StackingBox = {
+            position: [newStartX, boxes.length * BOX_HEIGHT, newStartZ],
+            args: [newWidth, BOX_HEIGHT, newLength],
+            color: `hsl(${Math.random() * 360}, 100%, 50%)`,
+        };
+
+        dispatch({ type: 'STACK_NEW_BOX', payload: { newBox } });
+    }, [boxes]);
 
-  useEffect(() => {
-    if (!direction) {
-      chooseRandomDirection(setDirection);
-    }
-  }, []);
+    useEffect(() => {
+        dispatch({ type: 'START_GAME', payload: { initialBoxes: initialBoxes } });
+    }, []);
 
-  useFrame(() => {
-    if (direction) moveInDirection[direction](movingBox.current, setDirection);
-  });
-  return { movingBox, lastBox, boxes, movingBoxDimesions };
+    useFrame(() => {
+        if (spacePressed) {
+            if (!wasPressed) {
+                if (direction === DIRECTION.NONE) {
+                    dispatch({ type: 'START_GAME', payload: { initialBoxes: initialBoxes } });
+                    setWasPressed(true);
+                } else {
+                    stackNewBox();
+                    setWasPressed(true);
+                }
+            }
+        } else {
+            if (wasPressed) setWasPressed(false);
+        }
+
+        if (direction !== DIRECTION.NONE) moveInDirection[direction](movingBox.current, dispatch);
+    });
```

Comment:
This will be cleaned up with subscription instead of always checking for it. https://github.com/pmndrs/drei/blob/0bfa86ddf8ba13dbe6a61dab6408d86b273316ec/README.md#keyboardcontrols